File: packs/dsa-linked-lists-py/challenges/02-middle-of-linked-list.test.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def to_list(head):
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result


def from_list(lst):
    dummy = ListNode(0)
    curr = dummy
    for val in lst:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next


from solution import middle_node


def test_middle_odd():
    head = from_list([1, 2, 3, 4, 5])
    result = middle_node(head)
    assert result.val == 3


def test_middle_even():
    head = from_list([1, 2, 3, 4, 5, 6])
    result = middle_node(head)
    assert result.val == 4


def test_middle_single():
    head = from_list([1])
    result = middle_node(head)
    assert result.val == 1


def test_middle_two():
    head = from_list([1, 2])
    result = middle_node(head)
    assert result.val == 2

File: packs/dsa-linked-lists-ts/challenges/02-middle-of-linked-list.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function fromList(arr: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let curr = dummy;
  for (const val of arr) {
    curr.next = new ListNode(val);
    curr = curr.next;
  }
  return dummy.next;
}

const CHALLENGE = join(import.meta.dirname, '02-middle-of-linked-list.json');

describe('Middle of Linked List', () => {
  let middleNode: (head: ListNode | null) => ListNode | null;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ middleNode: typeof middleNode }>(
      join(tmp.tmpDir, 'solution.ts'),
    );
    middleNode = mod.middleNode;
  });

  afterAll(() => cleanup());

  it('returns the middle node for an odd-length list', () => {
    const result = middleNode(fromList([1, 2, 3, 4, 5]));
    expect(result!.val).toBe(3);
  });

  it('returns the second middle for an even-length list', () => {
    const result = middleNode(fromList([1, 2, 3, 4, 5, 6]));
    expect(result!.val).toBe(4);
  });

  it('returns the only node for a single-element list', () => {
    const result = middleNode(fromList([1]));
    expect(result!.val).toBe(1);
  });

  it('returns the second node for a two-element list', () => {
    const result = middleNode(fromList([1, 2]));
    expect(result!.val).toBe(2);
  });
});
